### PR TITLE
[bugfix] Combine access and job options for slurm '--constraint'

### LINKED ISSUE #1348: Passing a `--constraint` job option should be combined with the  `access` one for Slurm
If a `--contraint=A` option is used in the `access` field of the `SystemPartition` then a corresponding line is emitted inside the sbatch script.
Passing additionally a constraint in the `options` field of the `Job` class, such as  `--contstraint=B` emits a single line in the sbatch script after the line corresponding to `--constraint=A`.
Therefore, slurm takes into account only the `B` constraint.
The two constraint options should be combined and passed in a single line as: `--constraint=A,B`

### PR DESCRIPTION
**Pull request by teojgo**
* The `--constraint` options passed via `access` and `job.options`
  are combined in the resulting job script.

* Add unittests to test the expected behavior.

Fixes #1348

### CODE REVIEW

File: reframe/core/schedulers/slurm.py [hunk 1]
```diff
@@ -195,15 +195,34 @@ def emit_preamble(self, job):
...
+                self._format_option(','.join(constraints), '--constraint={0}'))
 
         preamble.append(self._format_option(hint, '--hint={0}'))
         prefix_patt = re.compile(r'(#\w+)')
         for opt in job.options:
-            if not prefix_patt.match(opt):
-                preamble.append('%s %s' % (self._prefix, opt))
-            else:
-                preamble.append(opt)
+            if not opt.strip().startswith(('-C', '--constraint')):
```
Comment (ekouts) [inline reframe/core/schedulers/slurm.py:221]: I am not sure about this, but shouldn't you also check whether it starts with `#SBATCH -C` or `#SBATCH --constraint`?

Comment (teojgo) [reply reframe/core/schedulers/slurm.py:221]: No, because the `#SBATCH` prefix is added by ReFrame

File: reframe/core/schedulers/slurm.py [hunk 2]
```diff
@@ -195,15 +195,34 @@ def emit_preamble(self, job):
...
 
         for opt in job.sched_access:
-            preamble.append('%s %s' % (self._prefix, opt))
+            if not opt.strip().startswith(('-C', '--constraint')):
+                preamble.append('%s %s' % (self._prefix, opt))
+
+        constraints = []
+        option_parser = ArgumentParser()
+        option_parser.add_argument('-C', '--constraint')
+        parsed_args, _ = option_parser.parse_known_args(job.sched_access)
```
Comment (vkarak) [inline reframe/core/schedulers/slurm.py:204]:
```suggestion
        parsed_options, _ = constraint_parser.parse_known_args(job.sched_access)
```

File: reframe/core/schedulers/slurm.py [hunk 3]
```diff
@@ -195,15 +195,34 @@ def emit_preamble(self, job):
...
+        if parsed_args.constraint:
+            constraints += [parsed_args.constraint]
```
Comment (vkarak) [inline reframe/core/schedulers/slurm.py:206]:
```suggestion
            constraints.append(parsed_options.constraint)
```

File: reframe/core/schedulers/slurm.py [hunk 4]
```diff
@@ -195,15 +195,34 @@ def emit_preamble(self, job):
...
+
+        # NOTE: Here last of the passed --constraint job options is taken
+        # into account in order to respect the behavior of slurm.
+        parsed_args, _ = option_parser.parse_known_args(job.options)
+        if parsed_args.constraint:
+            constraints += [parsed_args.constraint]
```
Comment (vkarak) [inline reframe/core/schedulers/slurm.py:212]: Same comments as above.

File: reframe/core/schedulers/slurm.py [hunk 5]
```diff
@@ -195,15 +195,34 @@ def emit_preamble(self, job):
...
+
+        if constraints:
+            preamble.append(
+                self._format_option(','.join(constraints), '--constraint={0}'))
```
Comment (vkarak) [inline reframe/core/schedulers/slurm.py:216]: Move the closing parenthesis to the next line.

File: reframe/core/schedulers/slurm.py (hunk 1 above)
Comment (vkarak) [inline reframe/core/schedulers/slurm.py:221]:
```suggestion
            if opt.strip().startswith(('-C', '--constraint')):
                # Constraints are already processed
                continue
            
            # all the rest unindented here
```

Comment (vkarak) [reply reframe/core/schedulers/slurm.py:221]: Also are you sure that passing a tuple to `startswith()` works as you intend here?? It's not documented.

File: unittests/test_schedulers.py [hunk 6]
```diff
@@ -370,6 +370,26 @@ def test_no_empty_lines_in_preamble(minimal_job):
         assert line != ''
 
 
+def test_combined_access_constraint(make_job, slurm_only):
+    job = make_job(sched_access=['--constraint=c1'])
+    job.options = ['--constraint=c2,c3']
```
Comment (vkarak) [inline unittests/test_schedulers.py:375]: Use the `-C` option here to exercise that path as well.

```suggestion
    job.options = ['-C c2,c3']
```

File: unittests/test_schedulers.py [hunk 7]
```diff
@@ -370,6 +370,26 @@ def test_no_empty_lines_in_preamble(minimal_job):
...
+    prepare_job(job)
+    with open(job.script_filename) as fp:
+        assert re.search(r'--constraint=c1,c2,c3$', fp.read(), re.MULTILINE)
+        assert re.search(r'--constraint=(c1|c2,c3)$', fp.read(),
+                         re.MULTILINE) is None
```
Comment (vkarak) [inline unittests/test_schedulers.py:380]:
```suggestion
        assert re.search(r'(?m)--constraint=c1,c2,c3$', fp.read())
        assert re.search(r'(?m)--constraint=(c1|c2,c3)$', fp.read()) is None
```

File: unittests/test_schedulers.py [hunk 8]
```diff
@@ -370,6 +370,26 @@ def test_no_empty_lines_in_preamble(minimal_job):
...
+
+
+def test_combined_access_multiple_constraints(make_job, slurm_only):
+    job = make_job(sched_access=['--constraint=c1'])
+    job.options = ['--constraint=c2', '--constraint=c3']
```
Comment (vkarak) [inline unittests/test_schedulers.py:385]:
```suggestion
    job.options = ['--constraint=c2', '-C c3']
```

File: unittests/test_schedulers.py [hunk 9]
```diff
@@ -370,6 +370,26 @@ def test_no_empty_lines_in_preamble(minimal_job):
...
+    prepare_job(job)
+    with open(job.script_filename) as fp:
+        assert re.search(r'--constraint=c1,c3$', fp.read(), re.MULTILINE)
+        assert re.search(r'--constraint=(c1|c2|c3)$', fp.read(),
+                         re.MULTILINE) is None
```
Comment (vkarak) [inline unittests/test_schedulers.py:390]: See my comments above.

File: reframe/core/schedulers/slurm.py (hunk 1 above)
Comment (teojgo) [reply reframe/core/schedulers/slurm.py:221]: Yes, I did not know either, it is mentioned here: https://docs.python.org/3/library/stdtypes.html#str.startswith that the prefix can be a tuple of prefixes to try.

**Comment by teojgo** [comment]
@vkarak in which section do you think I should add the a documentation note about this?

**Comment by vkarak** [comment]
@teojgo The point that @ekouts suggested above should be ok.

File: docs/manpage.rst [hunk 10]
```diff
@@ -341,6 +341,10 @@ Options controlling job submission
    If ``key`` starts with ``-`` or ``#``, the option will be passed verbatim to the job script.
```
Comment (vkarak) [inline docs/manpage.rst:341]: This sentence is also not accurate.

**Pull request merged**